csharp/src/Ice/Communicator-OutgoingConnectionFactory.cs

```diff
@@ -0,0 +1,210 @@
+// Copyright (c) ZeroC, Inc. All rights reserved.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace ZeroC.Ice
+{
+    public sealed partial class Communicator
+    {
+        private readonly Dictionary<(Endpoint, object?), List<Connection>> _outgoingConnections =
+            new(ConnectionMapKeyComparer.Instance);
+        private readonly Dictionary<(Endpoint, object?), Task<Connection>> _pendingOutgoingConnections =
+            new(ConnectionMapKeyComparer.Instance);
+        // We keep a map of the endpoints that recently resulted in a failure while establishing a connection. This is
+        // used to influence the selection of endpoints when creating new connections. Endpoints with recent failures
+        // are tried last.
+        private readonly ConcurrentDictionary<Endpoint, DateTime> _transportFailures = new();
```

Comment:
This dictionary should use `EndpointComparer.Equivalent` like the other dictionaries.

Reply:
We need a separate comparer as this map doesn't include the label